source/common/upstream/transport_socket_match_impl.cc

```diff
@@ -50,15 +50,28 @@ TransportSocketMatchStats TransportSocketMatcherImpl::generateStats(const std::s
   return {ALL_TRANSPORT_SOCKET_MATCH_STATS(POOL_COUNTER_PREFIX(stats_scope_, prefix))};
 }
 
-TransportSocketMatcher::MatchData
-TransportSocketMatcherImpl::resolve(const envoy::config::core::v3::Metadata* metadata) const {
+TransportSocketMatcher::MatchData TransportSocketMatcherImpl::resolve(
+    const envoy::config::core::v3::Metadata* endpoint_metadata,
+    const envoy::config::core::v3::Metadata* locality_metadata) const {
+  // We want to check for a match in the endpoint metadata first, since that will always take
+  // precedence for transport socket matching.
   for (const auto& match : matches_) {
     if (Config::Metadata::metadataLabelMatch(
-            match.label_set, metadata,
+            match.label_set, endpoint_metadata,
             Envoy::Config::MetadataFilters::get().ENVOY_TRANSPORT_SOCKET_MATCH, false)) {
       return {*match.factory, match.stats, match.name};
     }
   }
+
+  // If we didn't match on any endpoint metadata, let's try for the locality.
```

Comment:
I'm good with this, but there's an implicit assumption that either endpoint and locality metadata are namespaced appropriately or mutually exclusive. It's probably true in reality, but maybe worth pointing out?

Reply:
By "namespaced appropriately" do you mean that the locality metadata would apply to all the endpoints it contains? If so, I assumed that the proto field's comment covers it, but I can mention it again here no problem. 

Correct me if I'm misunderstanding what you meant.

Reply:
As per IM, just an observation on API ergonomics, by mixing the two metadata sources with a common key space, it is a little less explicit what is going on, but it's fine.